File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/model/database/SetDataSource.java
package com.dbottillo.mtgsearchfree.model.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.dbottillo.mtgsearchfree.model.MTGSet;
import com.dbottillo.mtgsearchfree.util.LOG;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class SetDataSource {

    public static final String TABLE = "MTGSet";

    private SQLiteDatabase database;

    private enum COLUMNS {
        NAME("name", "TEXT"),
        CODE("code", "TEXT");

        private String name;
        private String type;

        COLUMNS(String name, String type) {
            this.name = name;
            this.type = type;
        }

        public String getName() {
            return name;
        }
    }

    public SetDataSource(SQLiteDatabase database) {
        this.database = database;
    }

    public static String generateCreateTable() {
        StringBuilder builder = new StringBuilder("CREATE TABLE IF NOT EXISTS ");
        builder.append(TABLE).append(" (_id INTEGER PRIMARY KEY, ");
        for (COLUMNS column : COLUMNS.values()) {
            builder.append(column.name).append(' ').append(column.type);
            if (column != COLUMNS.CODE) {
                builder.append(',');
            }
        }
        return builder.append(')').toString();
    }

    long saveSet(MTGSet set) {
        ContentValues values = new ContentValues();
        if (set.getId() > -1) {
            values.put("_id", set.getId());
        }
        values.put(COLUMNS.NAME.getName(), set.getName());
        values.put(COLUMNS.CODE.getName(), set.getCode());
        return database.insertWithOnConflict(TABLE, null, values, SQLiteDatabase.CONFLICT_REPLACE);
    }

    public List<MTGSet> getSets() {
        String query = "SELECT * FROM " + TABLE;
        Cursor cursor = database.rawQuery(query, null);
        ArrayList<MTGSet> sets = new ArrayList<>();
        if (cursor.moveToFirst()) {
            while (!cursor.isAfterLast()) {
                sets.add(fromCursor(cursor));
                cursor.moveToNext();
            }
        }
        LOG.query(query);
        cursor.close();
        return sets;
    }

    void removeSet(long id) {
        String[] args = new String[]{id + ""};
        String query = "DELETE FROM " + TABLE + " where _id=? ";
        Cursor cursor = database.rawQuery(query, args);
        cursor.moveToFirst();
        cursor.close();
        LOG.query(query, args);
    }

    public MTGSet fromCursor(Cursor cursor) {
        MTGSet set = new MTGSet(cursor.getInt(cursor.getColumnIndex("_id")));
        set.setName(cursor.getString(cursor.getColumnIndex(COLUMNS.NAME.getName())));
        set.setCode(cursor.getString(cursor.getColumnIndex(COLUMNS.CODE.getName())));
        return set;
    }

    public ContentValues fromJSON(JSONObject object) throws JSONException {
        ContentValues values = new ContentValues();
        values.put(COLUMNS.CODE.getName(), object.getString("code"));
        values.put(COLUMNS.NAME.getName(), object.getString("name"));
        return values;
    }
}

File: MTGSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/model/database/SetDataSource.kt
package com.dbottillo.mtgsearchfree.model.database

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.dbottillo.mtgsearchfree.model.MTGSet
import com.dbottillo.mtgsearchfree.util.LOG
import org.json.JSONException
import org.json.JSONObject

class SetDataSource(private val database: SQLiteDatabase) {

    val sets: List<MTGSet>
        get() {
            val query = "SELECT * FROM $TABLE"
            val cursor = database.rawQuery(query, null)
            val sets = ArrayList<MTGSet>()
            if (cursor.moveToFirst()) {
                while (!cursor.isAfterLast) {
                    sets.add(fromCursor(cursor))
                    cursor.moveToNext()
                }
            }
            LOG.query(query)
            cursor.close()
            return sets
        }

    fun saveSet(set: MTGSet): Long {
        val values = ContentValues().apply {
            if (set.id > -1) {
                put("_id", set.id)
            }
            put("name", set.name)
            put("code", set.code)
        }
        return database.insertWithOnConflict(TABLE, null, values, SQLiteDatabase.CONFLICT_REPLACE)
    }

    fun removeSet(id: Long) {
        val args = arrayOf(id.toString() + "")
        val query = "DELETE FROM $TABLE where _id=? "
        val cursor = database.rawQuery(query, args)
        cursor.moveToFirst()
        cursor.close()
        LOG.query(query, *args)
    }

    fun fromCursor(cursor: Cursor): MTGSet {
        return MTGSet(id = cursor.getIntFromColumn("_id"),
                name = cursor.getStringFromColumn("name"),
                code = cursor.getStringFromColumn("code"))
    }

    @Throws(JSONException::class)
    fun fromJSON(jsonObject: JSONObject): ContentValues {
        return ContentValues().apply {
            fromJson("name", jsonObject)
            fromJson("code", jsonObject)
        }
    }

    companion object {

        const val TABLE = "MTGSet"

        fun generateCreateTable(): String {
            val builder = StringBuilder("CREATE TABLE IF NOT EXISTS ")
            builder.append(TABLE).append(" (_id INTEGER PRIMARY KEY, ")
            builder.add(name = "name", type = "TEXT")
            builder.add(name = "code", type = "TEXT", last = true)
            return builder.append(')').toString()
        }
    }
}

fun ContentValues.fromJson(name: String, jsonObject: JSONObject) {
    put(name, jsonObject.getString(name))
}

fun StringBuilder.add(name: String, type: String, last: Boolean = false) {
    append(name).append(' ').append(type)
    if (!last) {
        append(",")
    }
}

fun Cursor.getIntFromColumn(column: String): Int {
    return getInt(getColumnIndex(column))
}

fun Cursor.getStringFromColumn(column: String): String {
    return getString(getColumnIndex(column))
}